src/services/poolTogether.js

```diff
@@ -0,0 +1,416 @@
+    .estimateAccountTransaction(estimateTransaction)
+    .then(result => buildTxFeeInfo(result, useGasToken))
+    .catch(() => null);
+
+  if (!estimated) {
+    return defaultResponse;
+  }
+
+  return estimated;
+};
+
+export async function getApproveFeeAndTransaction(symbol: string, useGasToken: boolean) {
+  const {
+    tokenContractAddress: contractAddress,
+    poolContractAddress,
+    unitType: decimals,
+    tokenABI,
+  } = getPoolTogetherTokenContract(symbol);
+  const transferMethod = tokenABI.find(item => item.name === 'approve');
+  const rawValue = 1000000000;
+  const valueToApprove = utils.parseUnits(rawValue.toString(), decimals);
+  const data = abi.encodeMethod(transferMethod, [poolContractAddress, valueToApprove]);
+  let transactionPayload = {
+    amount: 0,
+    to: contractAddress,
+    symbol,
+    contractAddress,
+    decimals,
+    data,
+    extra: {
+      poolTogetherApproval: {
+        symbol,
+      },
+    },
+  };
+
+  const { fee: txFeeInWei, gasToken } = await getSmartWalletTxFee(transactionPayload, useGasToken);
+  if (gasToken) {
+    transactionPayload = { ...transactionPayload, gasToken };
+  }
+
+  transactionPayload = { ...transactionPayload, txFeeInWei };
+
+  return {
+    gasToken,
+    txFeeInWei,
+    transactionPayload,
+  };
+}
+
+export const checkPoolAllowance = async (symbol: string, address: string): Promise<boolean> => {
+  const {
+    poolContractAddress,
+    tokenContractAddress: contractAddress,
+    tokenContract: contract,
+  } = getPoolTogetherTokenContract(symbol);
+  let hasAllowance = false;
+  try {
+    const allowanceResult = await contract.allowance(address, poolContractAddress);
+    if (allowanceResult) {
+      hasAllowance = allowanceResult.toString() !== '0';
+    }
+  } catch (e) {
+    reportLog('Error checking PoolTogether Allowance', {
+      address,
+      poolContractAddress,
+      contractAddress,
+      symbol,
+      message: e.message,
+    }, Sentry.Severity.Error);
+    hasAllowance = false;
+  }
+  return hasAllowance;
+};
+
+export async function getPurchaseTicketFeeAndTransaction(depositAmount: number, symbol: string, useGasToken: boolean) {
+  const {
+    poolAbi,
+    unitType: decimals,
+    poolContractAddress,
+  } = getPoolTogetherTokenContract(symbol);
+  const depositMethod = poolAbi.find(item => item.name === 'depositPool');
+  const valueToDeposit = utils.parseUnits(depositAmount.toString(), decimals);
+  const data = abi.encodeMethod(depositMethod, [valueToDeposit]);
+  let transactionPayload = {
+    amount: 0,
+    to: poolContractAddress,
+    symbol,
+    poolContractAddress,
+    decimals,
+    data,
+    extra: {
+      symbol,
+      amount: valueToDeposit,
+      decimals,
+    },
+    tag: POOLTOGETHER_DEPOSIT_TRANSACTION,
+  };
+
+  const { fee: txFeeInWei, gasToken } = await getSmartWalletTxFee(transactionPayload, useGasToken);
+  if (gasToken) {
+    transactionPayload = { ...transactionPayload, gasToken };
+  }
```

Comment:
What happens if we have an error with the gas estimation?

Reply:
an error will be thrown, logged and the users will have to try again for the fee

Reply:
Thanks for the explanation :)